irohad/torii/impl/command_service.cpp

```diff
@@ -155,140 +168,104 @@ namespace torii {
+    rxcpp::composite_subscription subscription;
 
-    log_->debug("StatusStream waiting finish, hash: {}", request_hash->hex());
+    auto hash = shared_model::crypto::Hash(request->tx_hash());
 
-    if (not*finished) {
-      resp = cache_->findItem(shared_model::crypto::Hash(request.tx_hash()));
-      if (not resp) {
-        log_->warn("StatusStream request processing timeout, hash: {}",
-                   request_hash->hex());
-        // TODO 05/03/2018 andrei IR-1046 Server-side shared model object
-        // factories with move semantics
-        auto resp_none = shared_model::proto::TransactionStatusBuilder()
-                             .txHash(*request_hash)
-                             .notReceived()
-                             .build();
-        response_writer.WriteLast(resp_none.getTransport(),
-                                  grpc::WriteOptions());
-      } else {
-        log_->debug(
-            "Tx processing was started but unfinished, awaiting more, hash: {}",
-            request_hash->hex());
-        /// We give it 2*proposal_delay time until timeout.
-        cv->wait_for(lock, 2 * proposal_delay_);
+    StatusStream(hash)
+        .map([](auto response) {
+          return std::static_pointer_cast<
+                     shared_model::proto::TransactionResponse>(response)
+              ->getTransport();
+        })
+        .lift<iroha::protocol::ToriiResponse>(
+            makeTimeout<iroha::protocol::ToriiResponse>(
+                [&](const auto &response) {
+                  return response.tx_status()
+                          == iroha::protocol::TxStatus::NOT_RECEIVED
+                      ? start_tx_processing_duration_
+                      : 2 * proposal_delay_;
+                },
+                current_thread))
+        .take_while([=](const auto &) { return not context->IsCancelled(); })
+        .subscribe(
+            subscription,
+            [&](iroha::protocol::ToriiResponse response) {
+              log_->debug("writing status");
+              response_writer->Write(response);
+            },
+            [&](std::exception_ptr ep) { log_->debug("processing timeout"); },
+            [&] { log_->debug("stream done"); });
 
-        /// status can be in the cache if it was finalized before we subscribed
-        if (not*finished) {
-          log_->debug("Transaction {} still not finished", request_hash->hex());
-          resp =
-              cache_->findItem(shared_model::crypto::Hash(request.tx_hash()));
-          log_->debug("Status of tx {} in cache: {}",
-                      request_hash->hex(),
-                      resp->tx_status());
-        }
-      }
-    } else {
-      log_->debug("StatusStream request processed successfully, hash: {}",
-                  request_hash->hex());
+    // run loop while subscription is active or there are pending events in the
+    // queue
+    while (subscription.is_subscribed() or not rl.empty()) {
```

Comment:
better to wrap this function with an inline method for the improving readability.